Create docs for VACUUM simulator (#187)
pganalyze/pganalyze-docs: vacuum-advisor/vacuum-simulator.mdx

---
title: 'VACUUM Simulator in Schema Statistics'
backlink_href: /docs/vacuum-advisor
backlink_title: 'VACUUM Advisor'
---

VACUUM Simulator lets you tweak autovacuum settings to learn the relation
between settings and autovacuum patterns.
It also helps you find better autovacuum settings for the table.

## Overview

Autovacuum can be triggered by the following 3 causes:

1. Triggered by dead rows
2. Triggered by freeze age
3. Triggered by inserts

Each cause has related configuration settings. By tweaking these settings, you
can see changes in the simulation data.
There is a panel for each cause, and each panel has the actual/current value and
graph section to show how it was in the selected time range, and simulation
value and graph section to show how it is based on the collected statistics and
the adjustable configuration settings.
You can also create a permalink of your current simulation and modified settings
to share with your team members.

### Limitation

* The current autovacuum count can be much bigger than the simulation count with
  frequently autovacuumed tables, since the simulation can only perform

>   autovacuum every 10 minutes.

Author: Uhmm yeah, I understand your point, though I don't have good alternative.

I think people can still use simulation for these tables to:

* understand what is causing so many vacuums
* find out how they can reduce autovacuum count

How does mentioning these sound?

>   autovacuum every 10 minutes.

Reviewer: Yeah, that sounds reasonable. Maybe something like
```
* The current autovacuum count can be much bigger than the simulation count with
  frequently autovacuumed tables, since the simulation can only perform
  autovacuum every 10 minutes. The simulation can still be useful to understand why
  so many vacuums are performed, and to explore ways to reduce autovacuum frequency.
```
?

This section also contains the graph of avoidable growth, calculated based on
the following factors:

* How many reusable rows are available (already VACUUMed rows)
* How many dead rows are there (to be VACUUMed rows)

Let's say there are 10 reusable rows and 20 dead rows. With this situation, if
30 rows are INSERTed:

1. Use 10 reusable rows
2. Use 20 totally new rows

However, if 20 dead rows were already VACUUMed at this point, the reusable rows
will be 30 rows and all INSERTs will use these reusable rows without acquiring
space for new rows. We'll consider these 20 dead rows as avoidable growth.
When the table is observing this pattern a lot, it indicates that autovacuums
are not running frequent enough. Try tweaking configuration settings to trigger
simulation autovacuums more and see if avoidable growth decreases.

There is a specific insight for this:
[Bloat - Insufficient VACUUM Frequency](/docs/checks/vacuum/insufficient_vacuum_frequency).

### Limitation

* The simulation data tends to trigger autovacuum earlier than the actual
  triggered timing, especially with tables with lots of UPDATEs.
  Occasionally, UPDATEs operations don't create dead rows, or even decrease dead

>   rows when "mini autovacuum" happened with these operations.

Author: I actually kinda forgot where I got this wording from, but I think I got it from our 5 mins postgres: https://pganalyze.com/blog/5mins-postgres-debug-UPDATE-bloated-tables-auto-explain-pageinspect#hot-pruning-in-postgres

> Postgres can actually do a “mini-VACUUM” of sorts, where it just vacuums the page that you're modifying, because it’s already modifying that page anyway.

I'll put the link to there

be triggered. These autovacuums are called anti-wraparound autovacuums.

With VACUUM Simulator, you can tweak `autovacuum_freeze_max_age` and
`autovacuum_multixact_freeze_max_age` in this section.
`autovacuum_multixact_freeze_max_age` can be only tweaked if multixact ID is
used. Lowering these configuration settings triggers autovacuums more, and
increasing these configuration settings will decrease autovacuums triggered by
freeze age.

### Limitation

* The simulation is using the data of average transactions (or multixacts) per
  second (tps).
  If the tps varies a lot depending on the time of the day or day of the week,
  the simulation may look a lot different than the actual data.
* Autovacuums triggered by other causes usually try to freeze old rows.
  The simulation also freezes them, however, there are cases that freezing won't
  happen when there are nothing can be frozen, or autovacuum is more focused on
  cleaning up dead rows.
  This can cause the transaction age to be reset even when it shouldn't be in
  the simulation, potentially cause the simulation to trigger less autovacuums
  by freeze age.

## Triggered by inserts

> Inserts autovacuum is introduced on Postgres 13.
> With an inserts only table, because it won't generate any dead rows, autovacuums
> won't be triggered by dead rows. It can still be triggered by freeze age, but
> depending on the database workload, it can be a long time until it's triggered.

Reviewer: I think we should lead with what insert-triggered autovacuum is (or why it exists), rather than what Postgres version introduced it. We should also point out why it's bad to wait a long time to freeze. And I think "append-only" is a more standard term than "inserts only"; we should probably stick to that.

```suggestion
Since an append-only table won't generate any dead rows, autovacuum
won't be triggered by dead rows. It can still be triggered by freeze age, but
depending on the database workload, it can be a long time until it's triggered.
This can turn normal incremental maintenance into a resource-intensive,
long-running task that can impact database performance. To counteract
this, since Postgres 13, autovacuum can be triggered by INSERTs.
```

> Inserts autovacuum is triggered to freeze old transaction IDs like freeze age

Reviewer: ```suggestion
INSERT-based autovacuum is triggered to freeze old transaction IDs like freeze age
```
or
```suggestion
Autovacuum based on inserts is triggered to freeze old transaction IDs like freeze age
```

autovacuum, but more importantly, it is triggered to update a visibility map.
Each table has a visibility map to keep track of which pages contain only tuples

> that are known to be visible to all active transactions. For the case like an
> inserts only table, all the pages of the table should be visible here and less
> likely to be not visible moving forward (will be flagged as not visible if there
> is any write operation on that page).

Reviewer: ```suggestion
that are known to be visible to all active transactions. For a case like an
append-only table, all pages of the table should be all-visible, and are
likely to stay all-visible moving forward (they will be flagged as not all-visible if there
is any write operation on that page).
```
I think it's important to talk about "all-visible" specifically, not just "visible", to keep things clear.

However, Postgres doesn't know about it until VACUUM runs to set a bit (flag).

> Updating the visibility map will help to spend less time for any subsequent
> VACUUMs as these pages don't need to be VACUUMed and will be skipped. Updating

Reviewer: ```suggestion
Updating the visibility map will improve performance for any subsequent
VACUUMs as all-visible pages don't need to be VACUUMed and will be skipped. Updating
```
We're kind of blurring the all-visible/all-frozen distinction here, but that's probably okay given the level of detail. I'm kind of torn: on the one hand, I think trying to add in that info might make this too complicated, but on the other it feels like this could be misleading otherwise. Thoughts?

Author: Uhmmm good point. After re-reading code:

* freezing vacuum: can't skip all-visible pages (but may skip all-frozen pages)
* dead rows and inserts vacuum: can skip all-visible pages

So "all-visible pages don't need to be VACUUMed" is not 100% accurate. We could add "unless it's a freezing (aggressive) vacuum" (I've been avoiding to use "aggressive vacuum" but I think I should be using as it is commonly used) to make it more accurate without introducing the word of "all-frozen".

Alternatively, inserts vacuum likely flag the page as both all-visible and all-frozen, so we can say something like:

> Updating the visibility map helps index-only scans, as they will be able to
avoid reading the table data to check visibility, improving performance.
> During the VACUUM of an append-only table, it is often able to mark pages as all-frozen, in addition to all-visible, and it'll stay that way moving forward too. This will improve performance for any subsequent VACUUMs as all-visible and all-frozen pages don't need to be VACUUMed and will be skipped.

Thoughts?

Reviewer: Yeah, that makes sense. Just minor wording tweaks:

>Updating the visibility map helps index-only scans, as they will be able to
>avoid reading table data to check visibility, improving performance.
>During VACUUM of an append-only table, it is often able to mark pages as all-frozen, in addition to all-visible, and it will stay that way (since these rows will never be modified). This will improve performance for any subsequent VACUUMs as pages that are both all-visible and all-frozen don't need to be VACUUMed and will be skipped.

(The most important is rewording that last sentence to avoid ambiguity.)

> the visibility map also helps to use index-only scans, as this information will
> be used to determinate weather to use index-only scans or not.

Reviewer: I think index-only scans can still be used without visibility map information, they just have to do extra table reads, right?

```suggestion
the visibility map also helps index-only scans, as they will be able to avoid reading
the table data to check visibility, improving performance.
```

Reviewer: Yeah, I think you're right: the live rows fraction is taken into account when costing IOSes. I don't think we need to talk about that, though, and we can focus on general IOS performance.

![Triggered by inserts](vacuum_simulator_inserts.png)

When the number of inserted rows (since the last VACUUM) exceeds the following
threshold, inserts autovacuums will be triggered:

```
threshold = autovacuum_vacuum_insert_threshold + autovacuum_vacuum_insert_scale_factor * number of tuples (pg_class.reltuples, number of live rows)
```

With VACUUM Simulator, you can tweak `autovacuum_vacuum_insert_threshold` and
`autovacuum_vacuum_insert_scale_factor` in this section.

> Lowering these configuration settings triggers autovacuums more with some
> inserts only or inserts heavy table, though this rarely triggers autovacuums for

Reviewer: ```suggestion
Lowering these configuration settings triggers autovacuums more with
append-only or append-mostly tables, though this rarely triggers autovacuums for
```

other types of tables.

If you are observing too many inserts autovacuums with the table, you can make
`autovacuum_vacuum_insert_scale_factor` 0 and adjust
`autovacuum_vacuum_insert_threshold` to find a reasonable threshold for inserts
autovacuums.

> ### Limitation

Reviewer: ```suggestion
### Limitations
```

* This simulation is only available with servers running Postgres 13 or above.

> * With the simulation, inserts autovacuum freezes old transaction IDs as it can
>   do it.
>   However, in reality, it's less likely that inserts autovacuum will freeze old
>   transaction IDs.

Reviewer: ```suggestion
* With the simulation, INSERT-based autovacuum freezes old transaction IDs as it can.
  However, in reality, it's less likely that it will freeze old transaction IDs.
```

Why less likely? I think this is also something we should try to explain better.

Author: I'm actually removing this, I wasn't understanding this behavior well enough -- freezing was skipped because the current oldest unfrozen transaction ID age was under `vacuum_freeze_min_age`. I'm planning to make a logic change in the simulator so we should be able to remove this limitation.
